### PR TITLE
VZ-7915: A-La-Carte test pipeline (WIP)

### PR DESCRIPTION
**Pull request by saidu941**
This PR introduces a new pipeline: verrazzano-a-la-carte; the idea is to install none profile, followed by edge stack, followed by application stack and verify installation at each step

Summary of changes:
> New Jenkins pipeline
> New verrazzano CR with none profile
> Tweaks to verify-install tests

### CODE REVIEW

**Comment by kminder** [comment]
I know this is WIP but please fill out the PR description.

File: ci/a-la-carte/Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+    parameters {
+        choice(name: 'KUBERNETES_CLUSTER_VERSION',
+                description: 'Kubernetes Version for KinD Cluster',
+                // 1st choice is the default value
+                choices: [ '1.24', '1.23', '1.22', '1.21' ])
+        string(name: 'GIT_COMMIT_TO_USE',
+                defaultValue: 'NONE',
+                description: 'This is the full git commit hash from the source build to be used for all jobs',
+                trim: true)
+        string(name: 'Installation_Profile',
```
Comment (kminder) [inline ci/a-la-carte/Jenkinsfile:34]: Make upper case for consistency.

File: ci/a-la-carte/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+            }
+        }
+
+        stage('Install KIND cluster with None') {
+            environment {
+                KIND_KUBERNETES_CLUSTER_VERSION = "${params.KUBERNETES_CLUSTER_VERSION}"
+                OCI_OS_LOCATION = "ephemeral/${env.BRANCH_NAME}/${SHORT_COMMIT_HASH}"
+                REALM_USER_PASSWORD = credentials('todo-mysql-password')
+                REALM_NAME = 'test-realm'
+                INSTALLATION_PROFILE = 'none'
```
Comment (kminder) [inline ci/a-la-carte/Jenkinsfile:200]: Does this make the job param irrelevant?

File: ci/a-la-carte/Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+            }
+
+            steps {
+                sh """
+                    cd ${GO_REPO_PATH}/verrazzano
+                    export INSTALLATION_PROFILE="none"
+                    ci/scripts/prepare_jenkins_at_environment.sh "" "" "" none
```
Comment (kminder) [inline ci/a-la-carte/Jenkinsfile:207]: Is the none passed here the same as the none profile?  Should the env var (or job param) be used?

Comment (kminder) [reply ci/a-la-carte/Jenkinsfile:207]: Add comments for what the empty strings mean

File: ci/a-la-carte/Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+            }
+            post {
+                always {
+                    archiveArtifacts artifacts: '**/coverage.html,**/logs/*,**/test-cluster-snapshots/**', allowEmptyArchive: true
+                    junit testResults: '**/*test-result.xml', allowEmptyResults: true
+                }
+            }
+        }
+    }
+        post {
```
Comment (kminder) [inline ci/a-la-carte/Jenkinsfile:255]: Indentation needs fixing.

File: ci/a-la-carte/Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+                currentBuild.displayName = env.BUILD_NUMBER + ' : PERIODIC'
+             } else if (causeString.contains('verrazzano-flaky-tests')) {
+                currentBuild.displayName = env.BUILD_NUMBER + ' : FLAKY'
+            }
+        }
+    }
+    echo 'End setDisplayName'
+}
+
+def dumpVerrazzanoSystemPods() {
```
Comment (kminder) [inline ci/a-la-carte/Jenkinsfile:330]: Why are all of these defined in the Jenkins file.  Can this be a script or make target in the repo?

File: ci/scripts/prepare_jenkins_at_environment.sh [hunk 6]
```diff
@@ -158,7 +158,10 @@ if [ $? -ne 0 ]; then
 fi
 
 echo "Installing Verrazzano on Kind"
-if [ -f "$WORKSPACE/vz" ]; then
+if [[ -z "${INSTALLATION_PROFILE}" ]]; then
```
Comment (kminder) [inline ci/scripts/prepare_jenkins_at_environment.sh:161]: This logic looks suspect to me.  The original if was checking for the existence of vz but the new if checks something and then assumes that vz exists.

File: ci/a-la-carte/Jenkinsfile (hunk 1 above)
Comment (kminder) [reply ci/a-la-carte/Jenkinsfile:34]: Is this even something that is allowed to be changed?

File: ci/a-la-carte/Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+                description: 'Verrazzano installation profile',
+                trim: true)
+        string(name: 'VERRAZZANO_OPERATOR_IMAGE',
+                defaultValue: 'NONE',
+                description: 'Verrazzano platform operator image name (in ghcr.io repo).  If not specified, the operator.yaml from Verrazzano repo will be used to create Verrazzano platform operator',
+                trim: true)
+        booleanParam(description: 'Whether to capture full cluster snapshot on test failure', name: 'CAPTURE_FULL_CLUSTER', defaultValue: false)
+    }
+
+    // TODO cleanup unnecessary ones
```
Comment (kminder) [inline ci/a-la-carte/Jenkinsfile:45]: Please do cleanup everything that is unnecessary.

File: ci/a-la-carte/Jenkinsfile [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+                        echo "Downloading VZ CLI from object storage"
+                        oci --region us-phoenix-1 os object get --namespace ${OCI_OS_NAMESPACE} -bn ${OCI_OS_COMMIT_BUCKET} --name ephemeral/${env.BRANCH_NAME}/${SHORT_COMMIT_HASH}/${VZ_CLI_TARGZ} --file ${VZ_CLI_TARGZ}
+                        tar xzf ${VZ_CLI_TARGZ} -C ${GO_REPO_PATH}
+                        ${GO_REPO_PATH}/vz version
+                    """
+                }
+            }
+        }
+
+        stage('Install KIND cluster with None') {
```
Comment (kminder) [inline ci/a-la-carte/Jenkinsfile:194]: Install KinD cluster with none profile

File: ci/a-la-carte/Jenkinsfile (hunk 1 above)
Comment (michael-cico) [reply ci/a-la-carte/Jenkinsfile:34]: yes, I don't think this should be a parameter, make it an environment var.  Also, call that variable INSTALL_PROFILE as that's what the setup scripts look for.

File: ci/a-la-carte/Jenkinsfile (hunk 5 above)
Comment (michael-cico) [reply ci/a-la-carte/Jenkinsfile:301]: this is boilerplate from a copied Jenkinsfile

Comment (michael-cico) [reply ci/a-la-carte/Jenkinsfile:301]: Not that we can't do that, but probably not a priority at the moment

File: ci/a-la-carte/Jenkinsfile [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+            environment {
+                KIND_KUBERNETES_CLUSTER_VERSION = "${params.KUBERNETES_CLUSTER_VERSION}"
+                OCI_OS_LOCATION = "ephemeral/${env.BRANCH_NAME}/${SHORT_COMMIT_HASH}"
+                REALM_USER_PASSWORD = credentials('todo-mysql-password')
+                REALM_NAME = 'test-realm'
+            }
+            steps {
+                sh """
+                    cd ${GO_REPO_PATH}/verrazzano
+                    ci/scripts/prepare_jenkins_at_environment.sh ${params.CREATE_CLUSTER_USE_CALICO} ${params.WILDCARD_DNS_DOMAIN} ${params.USE_DB_FOR_GRAFANA}
```
Comment (michael-cico) [inline ci/a-la-carte/Jenkinsfile:202]: You're missing these params I think.  For now, just pass in `true nip.io false`, e.g.,

```
ci/scripts/prepare_jenkins_at_environment.sh true nip.io false
```

File: ci/a-la-carte/Jenkinsfile [hunk 10]
```diff
@@ -0,0 +1,414 @@
...
+        stage('Install KIND cluster with None profile') {
+            environment {
+                KIND_KUBERNETES_CLUSTER_VERSION = "${params.KUBERNETES_CLUSTER_VERSION}"
+                OCI_OS_LOCATION = "ephemeral/${env.BRANCH_NAME}/${SHORT_COMMIT_HASH}"
+                REALM_USER_PASSWORD = credentials('todo-mysql-password')
+                REALM_NAME = 'test-realm'
+            }
+            steps {
+                sh """
+                    cd ${GO_REPO_PATH}/verrazzano
```
Comment (michael-cico) [inline ci/a-la-carte/Jenkinsfile:201]: I think before this you need to copy the test install yaml to the location the script expects:

```
                                    cp ${env.INSTALL_CONFIG_FILE_KIND_TESTS} ${env.INSTALL_CONFIG_FILE_KIND}
```

Comment (michael-cico) [reply ci/a-la-carte/Jenkinsfile:201]: The `INSTALL_CONFIG_FILE_KIND` variable is looked for by the prepare/process scripts.

File: ci/a-la-carte/Jenkinsfile [hunk 11]
```diff
@@ -0,0 +1,414 @@
...
+        NETRC_FILE = credentials('netrc')
+        CLUSTER_NAME = 'verrazzano'
+        POST_DUMP_FAILED_FILE = "${WORKSPACE}/post_dump_failed_file.tmp"
+        TESTS_EXECUTED_FILE = "${WORKSPACE}/tests_executed_file.tmp"
+        KUBECONFIG = "${WORKSPACE}/test_kubeconfig"
+        VERRAZZANO_KUBECONFIG = "${KUBECONFIG}"
+        OCR_CREDS = credentials('ocr-pull-and-push-account')
+        OCR_REPO = 'container-registry.oracle.com'
+        IMAGE_PULL_SECRET = 'verrazzano-container-registry'
+        INSTALL_CONFIG_FILE_KIND_TESTS = "./tests/e2e/config/scripts/${params.CRD_API_VERSION}/install-verrazzano-kind.yaml"
```
Comment (michael-cico) [inline ci/a-la-carte/Jenkinsfile:64]: change `params.CRD_API_VERSION` to use `v1beta1`, e.g.

```
INSTALL_CONFIG_FILE_KIND_TESTS = "./tests/e2e/config/scripts/v1beta1/install-verrazzano-kind.yaml"
```

File: ci/a-la-carte/Jenkinsfile [hunk 12]
```diff
@@ -0,0 +1,507 @@
...
+                OCI_OS_LOCATION = "ephemeral/${env.BRANCH_NAME}/${SHORT_COMMIT_HASH}"
+                REALM_USER_PASSWORD = credentials('todo-mysql-password')
+                REALM_NAME = 'test-realm'
+            }
+            steps {
+                sh """
+                    cp ${env.INSTALL_CONFIG_FILE_KIND_TESTS} ${env.INSTALL_CONFIG_FILE_KIND}
+                    cp ${INSTALL_CONFIG_FILE_KIND} ${WORKSPACE}/appstack.yaml
+                    cd ${GO_REPO_PATH}/verrazzano
+                    yq -i eval '.spec.components.kubeStateMetrics.enabled = false' ${WORKSPACE}/appstack.yaml
```
Comment (michael-cico) [inline ci/a-la-carte/Jenkinsfile:277]: Same comment, let's use the `update` pattern in test code on a subsequent pass. Probably when we do the Prom/Prom-operator story.

File: ci/a-la-carte/Jenkinsfile [hunk 13]
```diff
@@ -0,0 +1,507 @@
...
+        }
+
+        stage('verify-install promstack') {
+            environment {
+                DUMP_DIRECTORY="${TEST_DUMP_ROOT}/verify-install/promstack"
+            }
+            steps {
+                runGinkgoRandomize('verify-install/promstack')
+            }
+            post {
```
Comment (mbenmclean) [inline ci/a-la-carte/Jenkinsfile:256]: do cluster dump on failure

File: ci/a-la-carte/Jenkinsfile (hunk 12 above)
Comment (mbenmclean) [inline ci/a-la-carte/Jenkinsfile:277]: same comment about not doing yq

File: ci/a-la-carte/Jenkinsfile [hunk 14]
```diff
@@ -0,0 +1,507 @@
...
+        }
+
+        stage('verify-infra vmi') {
+            environment {
+                DUMP_DIRECTORY="${TEST_DUMP_ROOT}/verify-infra/vmi"
+            }
+            steps {
+                runGinkgoRandomize('verify-infra/vmi')
+            }
+            post {
```
Comment (mbenmclean) [inline ci/a-la-carte/Jenkinsfile:320]: do cluster dump on failure

File: ci/a-la-carte/Jenkinsfile [hunk 15]
```diff
@@ -0,0 +1,507 @@
...
+                    yq -i eval '.spec.components.prometheusPushgateway.enabled = true' ${WORKSPACE}/edgestack.yaml
+                    kubectl apply -f ${WORKSPACE}/edgestack.yaml
+
+                    # wait for Verrazzano install to complete
+                    ${WORKSPACE}/tests/e2e/config/scripts/wait-for-verrazzano-install.sh
+                """
+            }
+
+            post {
+                failure {
```
Comment (michael-cico) [inline ci/a-la-carte/Jenkinsfile:240]: next pass, we need to integrate cluster snapshots on failure.

**Review by michael-cico — APPROVED** [review]

File: ci/a-la-carte/Jenkinsfile [hunk 16]
```diff
@@ -0,0 +1,507 @@
...
+                        trim: true)
+        string (name: 'VERRAZZANO_OPERATOR_IMAGE',
+                        defaultValue: 'NONE',
+                        description: 'Verrazzano platform operator image name (in ghcr.io repo).  If not specified, the operator.yaml from Verrazzano repo will be used to create Verrazzano platform operator',
+                trim: true)
+        booleanParam (description: 'Whether to capture full cluster snapshot on test failure', name: 'CAPTURE_FULL_CLUSTER', defaultValue: false)
+    }
+
+    environment {
+        DOCKER_PLATFORM_CI_IMAGE_NAME = 'verrazzano-platform-operator-jenkins'
```
[the remark this answers is not in the document]
Comment (saidu941) [reply ci/a-la-carte/Jenkinsfile:42]: I don't see them used anywhere in the jenkinsfile, I can cleanup this.
This was based of KIND Jenkinsfile

File: ci/a-la-carte/Jenkinsfile [hunk 17]
```diff
@@ -0,0 +1,507 @@
...
+                OCI_OS_LOCATION = "ephemeral/${env.BRANCH_NAME}/${SHORT_COMMIT_HASH}"
+                REALM_USER_PASSWORD = credentials('todo-mysql-password')
+                REALM_NAME = 'test-realm'
+            }
+            steps {
+                sh """
+                    cp ${env.INSTALL_CONFIG_FILE_KIND_TESTS} ${env.INSTALL_CONFIG_FILE_KIND}
+                    cp ${INSTALL_CONFIG_FILE_KIND} ${WORKSPACE}/edgestack.yaml
+                    cd ${GO_REPO_PATH}/verrazzano
+                    yq -i eval '.spec.components.kubeStateMetrics.enabled = true' ${WORKSPACE}/edgestack.yaml
```
[the remark this answers is not in the document]
Comment (michael-cico) [reply ci/a-la-carte/Jenkinsfile:221]: We will move to the updater API I think in a subsequent pass.

File: ci/a-la-carte/Jenkinsfile [hunk 18]
```diff
@@ -0,0 +1,507 @@
...
+                    cp ${env.INSTALL_CONFIG_FILE_KIND_TESTS} ${env.INSTALL_CONFIG_FILE_KIND}
+                    cd ${GO_REPO_PATH}/verrazzano
+                    ci/scripts/prepare_jenkins_at_environment.sh true nip.io false
+
+                    # wait for Verrazzano install to complete
+                    ${WORKSPACE}/tests/e2e/config/scripts/wait-for-verrazzano-install.sh
+                """
+            }
+
+            post {
```
[the remark this answers is not in the document]
Comment (michael-cico) [reply ci/a-la-carte/Jenkinsfile:198]: Do as a follow-up PR.

File: ci/a-la-carte/Jenkinsfile (hunk 14 above)
Comment (michael-cico) [reply ci/a-la-carte/Jenkinsfile:323]: The `DUMP_DIRECTORY` should handle that right?  I believe the dump is baked into the test code?

**Review by michael-cico — APPROVED** [review]

**Pull request merged**